datafusion/physical-plan/src/joins/hash_join/shared_bounds.rs

```diff
@@ -165,149 +243,180 @@ impl SharedBoundsAccumulator {
+    /// * `Result<()>` - Ok if successful, Err if filter update failed or mode mismatch
+    pub(crate) async fn report_build_data(
         &self,
-        left_side_partition_id: usize,
-        partition_bounds: Option<Vec<ColumnBounds>>,
+        data: PartitionBuildDataReport,
     ) -> Result<()> {
-        // Store bounds in the accumulator - this runs once per partition
-        if let Some(bounds) = partition_bounds {
+        // Store data in the accumulator
+        {
             let mut guard = self.inner.lock();
 
-            let should_push = if let Some(last_bound) = guard.bounds.last() {
-                // In `PartitionMode::CollectLeft`, all streams on the left side share the same partition id (0).
-                // Since this function can be called multiple times for that same partition, we must deduplicate
-                // by checking against the last recorded bound.
-                last_bound.partition != left_side_partition_id
-            } else {
-                true
-            };
-
-            if should_push {
-                guard
-                    .bounds
-                    .push(PartitionBounds::new(left_side_partition_id, bounds));
+            match (data, &mut *guard) {
+                // Partitioned mode
+                (
+                    PartitionBuildDataReport::Partitioned {
+                        partition_id,
+                        bounds,
+                    },
+                    AccumulatedBuildData::Partitioned { partitions },
+                ) => {
+                    if let Some(bounds) = bounds {
+                        partitions[partition_id] = Some(PartitionedBuildData {
+                            partition_id,
+                            bounds,
+                        });
+                    }
+                }
+                // CollectLeft mode (store once, deduplicate across partitions)
+                (
+                    PartitionBuildDataReport::CollectLeft { bounds },
+                    AccumulatedBuildData::CollectLeft { data },
+                ) => {
+                    match (bounds, data) {
+                        (None, _) | (_, Some(_)) => {
+                            // No bounds reported or already reported; do nothing
+                        }
+                        (Some(new_bounds), data) => {
+                            // First report, store the bounds
+                            *data = Some(CollectLeftBuildData { bounds: new_bounds });
+                        }
+                    }
+                }
+                // Mismatched modes - should never happen
+                _ => {
+                    return datafusion_common::internal_err!(
+                        "Build data mode mismatch in report_build_data"
+                    );
+                }
             }
         }
 
+        // Wait for all partitions to report
         if self.barrier.wait().await.is_leader() {
```

Comment:
It would now be pretty easy to always update the filter and use `let fall_through_case = !self.barrier.wait().await.is_leader()` to set the default case for the `CASE` expression such that it's `true` if we haven't received data for all partitions and `false` if we have. That would allow immediate pushdown for partitions that have their build side complete before others.